### PR TITLE
Merge user-make-request into Master

### PR DESCRIPTION
**Pull request by sehoven**
Done code and XML as much as it can be for making a request. I put some comments at the top of MakeRequestActivity for some concerns I had while writing the code there that may or may not need to be addressed (I could just be overthinking everything and it's fine).

Blocks:
- Waiting for completed FareCalculator (using MockFareCalculator to generate random, fake data)
- Waiting for implementation of maps for the user to choose start and end locations on a map (placeholder button, fake data used, XML and code will likely change to accommodate our implementation)

### CODE REVIEW

**Review by Hreherch — CHANGES_REQUESTED** [review]
The description box might be too small. Or it allows me to enter too many lines. If it is one line, I expect hitting enter to hide the keyboard.

Why does it prompt me to "estimate the ride fare"? I feel like this button press is extraneous and maybe should be automatic? And the warning should maybe tell me that I "must enter the fare you are willing to pay" (or so)?
Unfortunately, even though I pressed the estimate button, it told me to estimate it? 

Should there be a visual component showing me the coordinates of my location that is more permanent than Toast? Do we need to have the ability to enter locations with numbers ourselves? 

Coolio otherwise 🌮 👍

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -49,8 +48,10 @@
         <activity
             android:name=".SearchMapsActivity"
             android:label="@string/title_activity_search_maps" />
-        <activity android:name=".SearchActivity"></activity>
->>>>>>> refs/remotes/origin/master
+        <activity android:name=".SearchActivity" />
+        >>>>>>> refs/remotes/origin/master
```
Comment (Hreherch) [inline app/src/main/AndroidManifest.xml]: What does line 52 mean?

File: app/src/main/java/comcmput301f16t01/github/carrier/MakeRequestActivity.java [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+those are necessary attributes for a Request to exist. I am assuming that if the
+user doesn't make any changes to the location, the fare estimate will be the same
+every time the method is called so maybe this is a non-issue but it just feels messy
+to have a method that does it, but here I have to call the method directly from its
+own class and then explicitly set the Request fare.
+
+I don't know if this is even a problem or if I am just overthinking things but
+whoever's reviewing my code give me your thoughts or ask me if my use of these
+classes and their methods doesn't make sense...
+ */
```
Comment (Hreherch) [inline app/src/main/java/comcmput301f16t01/github/carrier/MakeRequestActivity.java]: Fair. 
FairEstimate class will no longer be a dependency of Request in this case.

File: app/src/main/java/comcmput301f16t01/github/carrier/MakeRequestActivity.java [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_make_request);
+        setTitle("New Request");
+
+        activity.getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_PAN);
+
+        Button locationButton = (Button) findViewById(R.id.button_chooseLocation);
```
Comment (Hreherch) [inline app/src/main/java/comcmput301f16t01/github/carrier/MakeRequestActivity.java]: You can do this in XML.

android:onClick="functionName"

(same for the other two below)

File: app/src/main/java/comcmput301f16t01/github/carrier/MakeRequestActivity.java [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+     *      Pop up an AlertDialog to confirm that the user will cancel their request.
+     *      Return to MainActivity.
+     */
+    public void onBackPressed() {
+        AlertDialog.Builder adb = new AlertDialog.Builder(this);
+        adb.setTitle("Are you sure?");
+        adb.setMessage("Cancel request?");
+        adb.setCancelable(true);
+        final Activity activity = MakeRequestActivity.this;
+        // TODO semantics re: cancel, button text (is it confusing to have "cancel" and "cancel request"?)
```
Comment (Hreherch) [inline app/src/main/java/comcmput301f16t01/github/carrier/MakeRequestActivity.java:72]: Leave?
Are you sure? You will lose this request if you leave.
(Yes) (Stay | No | oh god )

File: app/src/main/java/comcmput301f16t01/github/carrier/MakeRequestActivity.java [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+            TextView fareTextView = (TextView) findViewById(R.id.textView_fareEstimate);
+            fareTextView.setText(formatFare(fareEstimate));
+
+            fareEstimated = fareEstimate;
+        }
+    }
+
+    /**
+     * Format the fare we get as a double as a string to be printed on the screen.
+     * This string will be preceded by a dollar sign (are we concerned about locale?)
```
Comment (Hreherch) [inline app/src/main/java/comcmput301f16t01/github/carrier/MakeRequestActivity.java:108]: Maybe consider locale.
http://stackoverflow.com/questions/7131922/how-to-format-a-float-value-with-the-device-currency-format
There is a built in class called NumberFormat, but I don't see this as a priority.

File: app/src/main/java/comcmput301f16t01/github/carrier/MakeRequestActivity.java [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+                request = new Request(user, start, end);
+            } else {
+                request = new Request(user, start, end, description);
+            }
+
+            request.setFare(fareEstimated);
+
+            rc.addRequest(request);
+
+            // TODO save the data (elasticsearch...will this be automatically done from within rc?)
```
Comment (Hreherch) [inline app/src/main/java/comcmput301f16t01/github/carrier/MakeRequestActivity.java]: Should be automatic in RC.

File: app/src/main/java/comcmput301f16t01/github/carrier/MakeRequestActivity.java [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+    public void submitRequest() {
+        RequestController rc = new RequestController();
+        UserController uc = new UserController();
+
+        if(start == null || end == null) {
+            Toast.makeText(activity, "You must first select a start and end location", Toast.LENGTH_SHORT).show();
+        } else if (fareEstimated != -1) {
+            Toast.makeText(activity, "You must first estimate the ride fare", Toast.LENGTH_SHORT).show();
+        } else {
+            User user = uc.getLoggedInUser();
```
Comment (Hreherch) [inline app/src/main/java/comcmput301f16t01/github/carrier/MakeRequestActivity.java]: These checks should be done within Request Controller. (I think this is an MVC thing, not sure)
That ensures a singular area for checks (if we ever have more than one area to create a request) :)
It also makes the view not responsible for things other than displaying and informing the controller of a potential request. 
Check out how I handle this sort of thing in the new user activity :)

File: app/src/main/AndroidManifest.xml (hunk 1 above)
Comment (sehoven) [reply app/src/main/AndroidManifest.xml]: I think Git put that in there...I'm not sure what that means either or where that came from...

File: app/src/main/java/comcmput301f16t01/github/carrier/MakeRequestActivity.java (hunk 3 above)
Comment (sehoven) [reply app/src/main/java/comcmput301f16t01/github/carrier/MakeRequestActivity.java]: Ah yes. Good observation. Thank you.

File: app/src/main/java/comcmput301f16t01/github/carrier/MakeRequestActivity.java (hunk 4 above)
Comment (sehoven) [reply app/src/main/java/comcmput301f16t01/github/carrier/MakeRequestActivity.java:72]: Changed the dialog to be more clear.

File: app/src/main/java/comcmput301f16t01/github/carrier/MakeRequestActivity.java (hunk 5 above)
Comment (sehoven) [reply app/src/main/java/comcmput301f16t01/github/carrier/MakeRequestActivity.java:108]: I agree that this is not a priority. I made an enhancement issue so we can address it if we have time. We could ask at our meeting today if this is something that we are expected to address.

File: app/src/main/java/comcmput301f16t01/github/carrier/MakeRequestActivity.java (hunk 7 above)
Comment (sehoven) [reply app/src/main/java/comcmput301f16t01/github/carrier/MakeRequestActivity.java]: Fixed this to handle requests in Request Controller.

File: app/src/main/java/comcmput301f16t01/github/carrier/MakeRequestActivity.java (hunk 6 above)
Comment (sehoven) [reply app/src/main/java/comcmput301f16t01/github/carrier/MakeRequestActivity.java]: Addressed this by adding an ElasticRequestController.

**Comment by sehoven** [comment]
> The description box might be too small. Or it allows me to enter too many lines. If it is one line, I expect hitting enter to hide the keyboard.

I made the description box bigger so it's obvious that it's multiline. I set the length of the description so they can enter as many lines up to the max character length, which we may want to change (I just chose 200 but that may be too big).

> Why does it prompt me to "estimate the ride fare"? I feel like this button press is extraneous and maybe should be automatic? And the warning should maybe tell me that I "must enter the fare you are willing to pay" (or so)?
> Unfortunately, even though I pressed the estimate button, it told me to estimate it?

I'm kind of confused about your feedback here so maybe explain to me what you mean in the lab today.

> Should there be a visual component showing me the coordinates of my location that is more permanent than Toast? Do we need to have the ability to enter locations with numbers ourselves?

The location button and the Toasts that come up are placeholders for now because we haven't figured out yet how we are dealing with the MapView and how much the MapActivity will be responsible for. Since this is dealing with a use case that is unrelated to creating a new request, I just made fake data to populate the request for now. We need to decide a) how the user selects the start and end locations on a map, b) how they are given the capability to see the locations they chose after they choose them.

**Review by Hreherch — APPROVED** [review]

**Comment by Hreherch** [comment]
Maybe remove line 52 in the manifest and fix the merge conflicts, but otherwise this is good for merging. 👍

**Pull request merged**